File: eios-go/internal/governance/audit.go
package governance

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"sync"
	"time"

	"eios/internal/domain"
)

// AuditLog is an append-only, hash-chained decision journal. Each entry's hash
// covers the previous hash, so any tampering breaks the chain (audit.ts).
type AuditLog struct {
	mu      sync.RWMutex
	events  []domain.AuditEvent
	nowFunc func() time.Time
}

func NewAuditLog() *AuditLog {
	return &AuditLog{nowFunc: time.Now}
}

func hashEntry(prevHash string, e domain.AuditEvent) string {
	h := sha256.Sum256([]byte(fmt.Sprintf("%s|%d|%s|%s|%s|%s|%s|%s",
		prevHash, e.Seq, e.Timestamp, e.Actor, e.ActorRole, e.Action, e.Resource, e.Detail)))
	return hex.EncodeToString(h[:])
}

// Append records an event and links it into the chain.
func (a *AuditLog) Append(actor, actorRole, action, resource, detail string) domain.AuditEvent {
	a.mu.Lock()
	defer a.mu.Unlock()
	prev := "genesis"
	if n := len(a.events); n > 0 {
		prev = a.events[n-1].Hash
	}
	e := domain.AuditEvent{
		Seq:       len(a.events) + 1,
		Timestamp: a.nowFunc().UTC().Format(time.RFC3339),
		Actor:     actor,
		ActorRole: actorRole,
		Action:    action,
		Resource:  resource,
		Detail:    detail,
		PrevHash:  prev,
	}
	e.Hash = hashEntry(prev, e)
	a.events = append(a.events, e)
	return e
}

func (a *AuditLog) Size() int { a.mu.RLock(); defer a.mu.RUnlock(); return len(a.events) }

// List returns the most recent `limit` events (newest last, as recorded).
func (a *AuditLog) List(limit int) []domain.AuditEvent {
	a.mu.RLock()
	defer a.mu.RUnlock()
	if limit <= 0 || limit > len(a.events) {
		limit = len(a.events)
	}
	start := len(a.events) - limit
	out := make([]domain.AuditEvent, limit)
	copy(out, a.events[start:])
	return out
}

type Integrity struct {
	Valid    bool `json:"valid"`
	BrokenAt int  `json:"brokenAt,omitempty"`
}

// Verify walks the chain and confirms every link.
func (a *AuditLog) Verify() Integrity {
	a.mu.RLock()
	defer a.mu.RUnlock()
	prev := "genesis"
	for _, e := range a.events {
		if e.PrevHash != prev || hashEntry(prev, e) != e.Hash {
			return Integrity{Valid: false, BrokenAt: e.Seq}
		}
		prev = e.Hash
	}
	return Integrity{Valid: true}
}

File: eios-prototype/server/src/governance/audit.ts
import { createHash } from 'node:crypto'
import type { AuditEvent, Role } from '../domain/types.js'
import { bus } from '../events/bus.js'

// Immutable, tamper-evident decision journal (Vol 7 Ch 10). Each entry hashes its
// content together with the previous entry's hash, forming a chain: any retroactive
// edit breaks verification. In production this is backed by append-only WORM storage.
class AuditLog {
  private events: AuditEvent[] = []
  private static readonly GENESIS = '0'.repeat(64)

  record(input: {
    actor: string
    actorRole: Role | 'system'
    action: string
    resource: string
    detail: string
  }): AuditEvent {
    const seq = this.events.length + 1
    const prevHash = this.events.at(-1)?.hash ?? AuditLog.GENESIS
    const timestamp = new Date().toISOString()
    const material = `${seq}|${timestamp}|${input.actor}|${input.actorRole}|${input.action}|${input.resource}|${input.detail}|${prevHash}`
    const hash = createHash('sha256').update(material).digest('hex')
    const event: AuditEvent = { seq, timestamp, ...input, prevHash, hash }
    this.events.push(event)
    return event
  }

  list(limit = 200): AuditEvent[] {
    // Return deep copies so consumers can never mutate the stored journal.
    return this.events.slice(-limit).reverse().map((e) => ({ ...e }))
  }

  /** Re-derive the chain to prove no entry was altered. */
  verify(): { valid: boolean; brokenAt?: number } {
    let prevHash = AuditLog.GENESIS
    for (const e of this.events) {
      const material = `${e.seq}|${e.timestamp}|${e.actor}|${e.actorRole}|${e.action}|${e.resource}|${e.detail}|${prevHash}`
      const expected = createHash('sha256').update(material).digest('hex')
      if (expected !== e.hash || e.prevHash !== prevHash) {
        return { valid: false, brokenAt: e.seq }
      }
      prevHash = e.hash
    }
    return { valid: true }
  }

  size(): number {
    return this.events.length
  }
}

export const auditLog = new AuditLog()

// Auto-journal every domain event so nothing EIOS does is invisible to audit.
bus.on('*', (e) => {
  const p = (e.payload ?? {}) as Record<string, unknown>
  auditLog.record({
    actor: (p.actor as string) ?? 'system',
    actorRole: (p.actorRole as Role) ?? 'system',
    action: e.type,
    resource: (p.resource as string) ?? '-',
    detail: (p.detail as string) ?? '',
  })
})
